File: packages/core/src/utils/deconstruction.ts
/**
 * 拆书工作台契约（G5/350 号，Phase C 第二大件首批；ANWA A6 拆书的确定性核心采纳）。
 *
 * 统一拆书面 = 把 chapter-analyzer 的逐章产物聚合为**可发布的拆书资产**：
 * 1. **四档人物档案** `deriveCharacterDossier`：brief（出场章）→ standard（+首末章/
 *    事件行）→ deep（+出场率/最长缺席）→ full（+全部证据原文）；
 * 2. **章节证据回溯** `buildEvidenceIndex`：角色 → [{章, 证据}] 索引
 *    （"这个角色在第几章做过什么"一键回溯）；
 * 3. **节奏/卖点统计** `analyzePacingStats`：章型序列占比 + 最长连续同型 + 强钩密度；
 * 4. **产物导出** `renderDeconstructionExport`：渲染为参考资料兼容 markdown
 *    （`## Extracted content` 标记，被 `reference-context.extractMaterialContent`
 *    消费——拆书产物可直接发布到参考资料供 G1 召回）。

 * 双端：`engine-rs/src/utils/deconstruction.rs` 1:1 镜像；共享向量
 * `src/__tests__/golden/deconstruction-vectors.json`。
 */

export const DECONSTRUCTION_DEPTHS = ["brief", "standard", "deep", "full"] as const;
export type DeconstructionDepth = (typeof DECONSTRUCTION_DEPTHS)[number];

/** 单章拆书证据（来自 chapter-analyzer 产物或素材分章）。 */
export interface DeconChapter {
  readonly chapter: number;
  /** 出场人物（已拆分）。 */
  readonly characters: ReadonlyArray<string>;
  /** 本章关键事件/梗概。 */
  readonly events: string;
  /** 章型（推进/铺垫/高潮…；可空）。 */
  readonly chapterType?: string;
  /** 伏笔动静（G10 强度口径可复用；可空）。 */
  readonly hookActivity?: string;
}

/** 证据回溯索引：角色 → 逐章证据。 */
export interface EvidenceEntry {
  readonly chapter: number;
  readonly evidence: string;
}

export interface EvidenceIndex {
  /** 角色 → 逐章证据（章号升序）。 */
  readonly byCharacter: Readonly<Record<string, EvidenceEntry[]>>;
  /** 全局章型序列（升序，仅含有章型的章）。 */
  readonly chapterTypes: ReadonlyArray<{ readonly chapter: number; readonly chapterType: string }>;
}

/** 证据索引构建（章号升序稳定；events 空白行不入索引）。 */
export function buildEvidenceIndex(chapters: ReadonlyArray<DeconChapter>): EvidenceIndex {
  const byCharacter: Record<string, EvidenceEntry[]> = {};
  const chapterTypes: Array<{ chapter: number; chapterType: string }> = [];
  for (const chapter of [...chapters].sort((a, b) => a.chapter - b.chapter)) {
    const evidence = chapter.events.trim();
    if (evidence) {
      for (const character of chapter.characters) {
        const name = character.trim();
        if (!name) continue;
        (byCharacter[name] ??= []).push({ chapter: chapter.chapter, evidence });
      }
    }
    if (chapter.chapterType?.trim()) {
      chapterTypes.push({ chapter: chapter.chapter, chapterType: chapter.chapterType.trim() });
    }
  }
  return { byCharacter, chapterTypes };
}

export interface CharacterDossier {
  readonly name: string;
  readonly appearances: ReadonlyArray<number>;
  readonly firstChapter: number;
  readonly lastChapter: number;
  /** 出场章数 / 书籍章数跨度（0–1，一位小数）。deep/full 档填充。 */
  readonly frequency?: number;
  /** 最长连续缺席章数。deep/full 档填充。 */
  readonly longestAbsence?: number;
  readonly evidence: ReadonlyArray<EvidenceEntry>;
}

/** 四档档案推导：depth 决定 detail 字段填充（brief 最薄，full 全量）。 */
export function deriveCharacterDossier(
  index: EvidenceIndex,
  name: string,
  depth: DeconstructionDepth,
  totalChapters?: number,
): CharacterDossier {
  const evidence = index.byCharacter[name] ?? [];
  const appearances = evidence.map((entry) => entry.chapter);
  const firstChapter = appearances[0] ?? 0;
  const lastChapter = appearances[appearances.length - 1] ?? 0;

  let longestAbsence = 0;
  if (appearances.length >= 2) {
    for (let i = 1; i < appearances.length; i++) {
      longestAbsence = Math.max(longestAbsence, appearances[i]! - appearances[i - 1]! - 1);
    }
  }
  const span = totalChapters ?? (lastChapter || 1);
  const frequency = span > 0 && appearances.length > 0
    ? Math.round((appearances.length / span) * 10) / 10
    : 0;

  return {
    name,
    appearances,
    firstChapter,
    lastChapter,
    ...(depth === "deep" || depth === "full" ? { frequency } : {}),
    ...(depth === "deep" || depth === "full" ? { longestAbsence } : {}),
    evidence:
      depth === "full"
        ? evidence
        : evidence.slice(0, depth === "brief" ? 0 : depth === "standard" ? 3 : 10),
  };
}

export interface PacingStats {
  /** 章型计数（含无章型章计入 "untyped"）。 */
  readonly counts: Readonly<Record<string, number>>;
  /** 最长连续同型章数与章型。 */
  readonly longestRun: { readonly chapterType: string; readonly length: number };
  /** 强钩密度：hookActivity 非空的章占比（0–1，一位小数）。 */
  readonly strongHookDensity: number;
}

/** 节奏/卖点统计：章型分布 + 最长连续同型 + 强钩密度（确定性口径）。 */
export function analyzePacingStats(chapters: ReadonlyArray<DeconChapter>): PacingStats {
  const counts: Record<string, number> = {};
  let longestType = "";
  let longestLength = 0;
  let runType = "";
  let runLength = 0;
  let strongChapters = 0;

  for (const chapter of chapters) {
    const type = chapter.chapterType?.trim() || "untyped";
    counts[type] = (counts[type] ?? 0) + 1;
    if (type === runType) {
      runLength += 1;
    } else {
      runType = type;
      runLength = 1;
    }
    if (runLength > longestLength) {
      longestLength = runLength;
      longestType = type;
    }
    if (chapter.hookActivity?.trim()) strongChapters += 1;
  }
  const total = chapters.length || 1;
  return {
    counts,
    longestRun: { chapterType: longestType, length: longestLength },
    strongHookDensity: Math.round((strongChapters / total) * 10) / 10,
  };
}

/** 导出渲染：参考资料兼容 markdown（`## Extracted content` 后为拆书正文）。 */
export function renderDeconstructionExport(
  dossiers: ReadonlyArray<CharacterDossier>,
  pacing: PacingStats,
  language: "zh" | "en" = "zh",
): string {
  const isEn = language === "en";
  const header = isEn
    ? "# Deconstruction Export"
    : "# 拆书产物（人物档案 / 节奏 / 卖点）";
  const dossierBlocks = dossiers.map((dossier) => {
    const lines = [
      `### ${dossier.name}`,
      isEn
        ? `- Chapters: ${dossier.appearances.join(", ") || "-"} | span ${dossier.firstChapter}–${dossier.lastChapter} | frequency ${dossier.frequency}`
        : `- 出场章：${dossier.appearances.join(", ") || "-"}｜跨度 ${dossier.firstChapter}–${dossier.lastChapter}｜出场率 ${dossier.frequency}`,
    ];
    for (const entry of dossier.evidence) {
      lines.push(`  - ch${entry.chapter}: ${entry.evidence}`);
    }
    return lines.join("\n");
  });
  const body = [
    header,
    "",
    "## Extracted content",
    "",
    isEn ? "### Character dossiers" : "### 人物档案",
    ...dossierBlocks,
    "",
    isEn
      ? `### Pacing: longest same-type run = ${pacing.longestRun.length} (${pacing.longestRun.chapterType}); strong-hook density ${pacing.strongHookDensity}`
      : `### 节奏：最长连续同型 ${pacing.longestRun.length} 章（${pacing.longestRun.chapterType}）；强钩密度 ${pacing.strongHookDensity}`,
  ];
  return body.join("\n");
}

/** 聚合产物（一次性产齐：索引 + 档案 + 节奏 + 可发布 markdown）。 */
export interface DeconstructionResult {
  readonly index: EvidenceIndex;
  readonly dossiers: ReadonlyArray<CharacterDossier>;
  readonly pacing: PacingStats;
  readonly markdown: string;
}

/**
 * 一括聚合（统一拆书面端点的核心调用）：按出场证据数取前 topCharacters
 * （默认 5）个角色的 full 档，加上节奏统计渲染为可发布 markdown。
 */
export function buildDeconstructionExport(params: {
  readonly chapters: ReadonlyArray<DeconChapter>;
  readonly depth?: DeconstructionDepth;
  readonly language?: "zh" | "en";
  readonly topCharacters?: number;
}): DeconstructionResult {
  const index = buildEvidenceIndex(params.chapters);
  const pacing = analyzePacingStats(params.chapters);
  const topCharacters = params.topCharacters ?? 5;
  const dossiers = Object.entries(index.byCharacter)
    .sort((a, b) => b[1].length - a[1].length || comparePlainDecon(a[0], b[0]))
    .slice(0, Math.max(1, topCharacters))
    .map(([name]) => deriveCharacterDossier(index, name, "full", undefined));
  const markdown = renderDeconstructionExport(dossiers, pacing, params.language ?? "zh");
  return { index, dossiers, pacing, markdown };
}

function comparePlainDecon(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

File: engine-rs/src/utils/deconstruction.rs
//! 拆书工作台契约（G5/350 号，Phase C 第二大件首批）。
//!
//! TS 真源：`packages/core/src/utils/deconstruction.ts`；共享向量：
//! `packages/core/src/__tests__/golden/deconstruction-vectors.json`
//! （差分测试 `tests/golden_deconstruction_diff.rs`）。
//!
//! 四档人物档案 / 章节证据回溯 / 节奏卖点统计 / 参考资料兼容导出，
//! 语义详见 TS 模块 doc。

use serde::Deserialize;
use serde::Serialize;

pub const DECONSTRUCTION_DEPTHS: [&str; 4] = ["brief", "standard", "deep", "full"];

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeconChapter {
    pub chapter: i64,
    #[serde(default)]
    pub characters: Vec<String>,
    #[serde(default)]
    pub events: String,
    #[serde(default)]
    pub chapter_type: Option<String>,
    #[serde(default)]
    pub hook_activity: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EvidenceEntry {
    pub chapter: i64,
    pub evidence: String,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EvidenceIndex {
    /// 角色首次出现序（对齐 TS 对象插入序；BTreeMap 会按码元重序破坏 duel 一致）。
    pub by_character: Vec<(String, Vec<EvidenceEntry>)>,
    pub chapter_types: Vec<(i64, String)>,
}

/// 证据索引构建（章号升序稳定；events 空白行不入索引）。
pub fn build_evidence_index(chapters: &[DeconChapter]) -> EvidenceIndex {
    let mut ordered: Vec<&DeconChapter> = chapters.iter().collect();
    ordered.sort_by_key(|chapter| chapter.chapter);

    let mut by_character: Vec<(String, Vec<EvidenceEntry>)> = Vec::new();
    let mut chapter_types: Vec<(i64, String)> = Vec::new();
    for chapter in ordered {
        let evidence = chapter.events.trim().to_string();
        if !evidence.is_empty() {
            for character in &chapter.characters {
                let name = character.trim().to_string();
                if name.is_empty() {
                    continue;
                }
                let slot = by_character
                    .iter_mut()
                    .find(|(existing, _)| existing == &name)
                    .map(|(_, entries)| entries);
                match slot {
                    Some(entries) => entries.push(EvidenceEntry {
                        chapter: chapter.chapter,
                        evidence: evidence.clone(),
                    }),
                    None => by_character.push((
                        name,
                        vec![EvidenceEntry {
                            chapter: chapter.chapter,
                            evidence: evidence.clone(),
                        }],
                    )),
                }
            }
        }
        if let Some(kind) = chapter.chapter_type.as_ref().map(|kind| kind.trim().to_string()).filter(|kind| !kind.is_empty()) {
            chapter_types.push((chapter.chapter, kind));
        }
    }
    EvidenceIndex { by_character, chapter_types }
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CharacterDossier {
    pub name: String,
    pub appearances: Vec<i64>,
    pub first_chapter: i64,
    pub last_chapter: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frequency: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub longest_absence: Option<i64>,
    pub evidence: Vec<EvidenceEntry>,
}

/// 四档档案推导：depth 决定 detail 字段填充与证据截断
/// （brief 0 条 / standard 3 条 / deep 10 条 / full 全量）。
pub fn derive_character_dossier(
    index: &EvidenceIndex,
    name: &str,
    depth: &str,
    total_chapters: Option<i64>,
) -> CharacterDossier {
    let empty: Vec<EvidenceEntry> = Vec::new();
    let evidence_full = index
        .by_character
        .iter()
        .find(|(existing, _)| existing == name)
        .map(|(_, entries)| entries)
        .unwrap_or(&empty);
    let appearances: Vec<i64> = evidence_full.iter().map(|entry| entry.chapter).collect();
    let first_chapter = appearances.first().copied().unwrap_or(0);
    let last_chapter = appearances.last().copied().unwrap_or(0);

    let mut longest_absence: i64 = 0;
    for pair in appearances.windows(2) {
        longest_absence = longest_absence.max(pair[1] - pair[0] - 1);
    }
    let span = total_chapters.unwrap_or(if last_chapter == 0 { 1 } else { last_chapter });
    let frequency = if span > 0 && !appearances.is_empty() {
        ((appearances.len() as f64 / span as f64) * 10.0).round() / 10.0
    } else {
        0.0
    };

    let cap = match depth {
        "brief" => 0usize,
        "standard" => 3,
        "deep" => 10,
        _ => usize::MAX,
    };
    CharacterDossier {
        name: name.to_string(),
        appearances,
        first_chapter,
        last_chapter,
        frequency: (depth == "deep" || depth == "full").then_some(frequency),
        longest_absence: (depth == "deep" || depth == "full").then_some(longest_absence),
        evidence: evidence_full.iter().take(cap).cloned().collect(),
    }
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LongestRun {
    pub chapter_type: String,
    pub length: i64,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PacingStats {
    pub counts: std::collections::BTreeMap<String, i64>,
    pub longest_run: LongestRun,
    pub strong_hook_density: f64,
}

/// 节奏/卖点统计：章型分布 + 最长连续同型 + 强钩密度（非空 hookActivity 占比）。
pub fn analyze_pacing_stats(chapters: &[DeconChapter]) -> PacingStats {
    let mut counts: std::collections::BTreeMap<String, i64> = std::collections::BTreeMap::new();
    let mut longest_type = String::new();
    let mut longest_length: i64 = 0;
    let mut run_type = String::new();
    let mut run_length: i64 = 0;
    let mut strong = 0i64;

    for chapter in chapters {
        let kind = chapter
            .chapter_type
            .as_ref()
            .map(|kind| kind.trim().to_string())
            .filter(|kind| !kind.is_empty())
            .unwrap_or_else(|| "untyped".to_string());
        *counts.entry(kind.clone()).or_insert(0) += 1;
        if kind == run_type {
            run_length += 1;
        } else {
            run_type = kind.clone();
            run_length = 1;
        }
        if run_length > longest_length {
            longest_length = run_length;
            longest_type = run_type.clone();
        }
        if !chapter.hook_activity.as_ref().map(|text| text.trim().is_empty()).unwrap_or(true) {
            strong += 1;
        }
    }
    let total = chapters.len().max(1) as f64;
    PacingStats {
        counts,
        longest_run: LongestRun { chapter_type: longest_type, length: longest_length },
        strong_hook_density: ((strong as f64 / total) * 10.0).round() / 10.0,
    }
}

/// 导出渲染：参考资料兼容 markdown（`## Extracted content` 后为拆书正文）。
pub fn render_deconstruction_export(
    dossiers: &[CharacterDossier],
    pacing: &PacingStats,
    language: &str,
) -> String {
    let is_en = language == "en";
    let header = if is_en {
        "# Deconstruction Export"
    } else {
        "# 拆书产物（人物档案 / 节奏 / 卖点）"
    };
    let mut body = vec![header.to_string(), String::new(), "## Extracted content".into()];
    body.push(if is_en { "### Character dossiers".into() } else { "### 人物档案".into() });
    for dossier in dossiers {
        body.push(format!("### {}", dossier.name));
        let appearances = if dossier.appearances.is_empty() {
            "-".to_string()
        } else {
            dossier
                .appearances
                .iter()
                .map(|chapter| chapter.to_string())
                .collect::<Vec<_>>()
                .join(", ")
        };
        body.push(if is_en {
            format!(
                "- Chapters: {appearances} | span {}–{} | frequency {}",
                dossier.first_chapter, dossier.last_chapter,
                dossier.frequency.unwrap_or(0.0)
            )
        } else {
            format!(
                "- 出场章：{appearances}｜跨度 {}–{}｜出场率 {}",
                dossier.first_chapter, dossier.last_chapter,
                dossier.frequency.unwrap_or(0.0)
            )
        });
        for entry in &dossier.evidence {
            body.push(format!("  - ch{}: {}", entry.chapter, entry.evidence));
        }
    }
    body.push(String::new());
    body.push(if is_en {
        format!(
            "### Pacing: longest same-type run = {} ({}); strong-hook density {}",
            pacing.longest_run.length, pacing.longest_run.chapter_type, pacing.strong_hook_density
        )
    } else {
        format!(
            "### 节奏：最长连续同型 {} 章（{}）；强钩密度 {}",
            pacing.longest_run.length, pacing.longest_run.chapter_type, pacing.strong_hook_density
        )
    });
    body.join("\n")
}

/// 机器可读契约（双端 golden 锁形状）。
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeconstructionContract {
    pub depths: Vec<&'static str>,
    pub evidence_caps: std::collections::BTreeMap<&'static str, usize>,
    pub export_marker: &'static str,
    pub consumed_by: &'static str,
}

pub fn deconstruction_contract() -> DeconstructionContract {
    let mut evidence_caps = std::collections::BTreeMap::new();
    evidence_caps.insert("brief", 0usize);
    evidence_caps.insert("standard", 3usize);
    evidence_caps.insert("deep", 10usize);
    DeconstructionContract {
        depths: DECONSTRUCTION_DEPTHS.to_vec(),
        evidence_caps,
        export_marker: "## Extracted content",
        consumed_by: "reference-context.extractMaterialContent",
    }
}

// ── 一括聚合（统一拆书面端点核心）──

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DeconstructionResult {
    pub index: EvidenceIndex,
    pub dossiers: Vec<CharacterDossier>,
    pub pacing: PacingStats,
    pub markdown: String,
}

/// 一括聚合：按出场证据数取前 top_characters 个角色的 full 档 + 节奏统计
/// 渲染为可发布 markdown。
pub fn build_deconstruction_export(
    chapters: &[DeconChapter],
    _depth: &str,
    language: &str,
    top_characters: Option<usize>,
) -> DeconstructionResult {
    let index = build_evidence_index(chapters);
    let pacing = analyze_pacing_stats(chapters);
    let top = top_characters.unwrap_or(5).max(1);
    let mut order: Vec<(String, usize)> = index
        .by_character
        .iter()
        .map(|(name, entries)| (name.clone(), entries.len()))
        .collect();
    order.sort_by(|a, b| b.1.cmp(&a.1).then(a.0.cmp(&b.0)));
    let dossiers: Vec<CharacterDossier> = order
        .into_iter()
        .take(top)
        .map(|(name, _)| derive_character_dossier(&index, &name, "full", None))
        .collect();
    let markdown = render_deconstruction_export(&dossiers, &pacing, language);
    DeconstructionResult { index, dossiers, pacing, markdown }
}
